subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <download_client.h>
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+#include <pm_config.h>
+#include <logging/log.h>
+#include <fota_download.h>
+
+LOG_MODULE_REGISTER(fota_download, CONFIG_FOTA_DOWNLOAD_LOG_LEVEL);
+
+static 		fota_download_callback_t callback;
+static struct	flash_img_context flash_img;
+static struct	download_client dfu;
+
+static int download_client_callback(const struct download_client_evt *event)
+{
+	int err;
+
+	__ASSERT(event != NULL, "invalid dfu object\n");
+
+	switch (event->id) {
+	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
+		size_t size;
+
+		err = download_client_file_size_get(&dfu, &size);
+		if (err != 0) {
+			LOG_ERR("download_client_file_size_get error %d",
+					err);
+			return err;
+		}
+		if (size > PM_MCUBOOT_SECONDARY_SIZE) {
+			LOG_ERR("Requested file too big to fit in flash\n");
+			return -EFBIG;
+		}
+
+		err = flash_img_buffered_write(&flash_img,
+				(u8_t *)event->fragment.buf,
+				event->fragment.len, false);
+		if (err != 0) {
+			LOG_ERR("flash_img_buffered_write error %d", err);
+			return err;
+		}
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DONE:
+		/* Write with 0 length to flush the write operation to flash. */
+		err = flash_img_buffered_write(&flash_img,
+				(u8_t *)event->fragment.buf,
+				0, true);
+		if (err != 0) {
+			LOG_ERR("flash_img_buffered_write error %d", err);
+			return err;
+		}
+
+		err = boot_request_upgrade(BOOT_UPGRADE_PERMANENT);
```

Comment:
How did you handle the erasing of the last page?

Reply:
That might be the missing point. I was not able to get this running successfully before pushing the change, but I wanted to push it to get reviews for the new variant. I assumed the change to fix would be small as this is already used by for instance `usb_dfu.c`.

Reply:
This is handled by the `flash_img` library.

As seen by the log:
```
[00:01:42.172,882] <inf> fota_flash_block: Erasing sector at offset 0x24000
[00:01:42.293,334] <inf> fota_flash_block: Erasing sector at offset 0x68000
```